### ISSUE TITLE
Feature Request: New Data Source `tfe_current_run`

### ISSUE THREAD

**Issue by joakimhellum** [hashicorp/terraform-provider-tfe#202]
**Why is this data source needed?**

We have been using a community Terraform provider for Git to get the current commit sha and tag for the repository, which we use to tag resources, for example:
```hcl

data "git_repository" "example" {
  path = path.module
}

resource "azurerm_resource_group" "example" {
  ...

  tags = {
    commit = substr(data.git_repository.example.commit_sha, 0, 7)
    tag    = data.git_repository.example.tag
  }
}
```

When Terraform uploads configuration versions to Terraform Cloud it excludes the `.git` folder by default.
Having a `tfe_current_run` data source which reads that information from the run environment could solve this without needing to upload the `.git` folder.  It also allows us to get the current workspace (#187)

Potential Terraform configuration:
```hcl
data "tfe_current_run" "example" {}

output "workspace_name" {
  value = data.tfe_current_run.example.workspace_name
}

output "commit_sha" {
  value = data.tfe_current_run.example.git_commit_sha
}

...
```

References:
https://www.terraform.io/docs/cloud/run/run-environment.html#environment-variables

I'm happy to create a pull request for this if you think it's reasonable.

**Comment by A30004028** [comment]
Totally agree, I am looking into using the tfe_run_trigger resource for triggering downstream events on child workspaces.

I'd like to associate the current in-use workspace to these created workspaces by way of tfe_run_trigger, but I cant see anyway to populate the value of the sourceable_id without hardcoded the current workspace name or id.

It'd be good to be able to call a data source against the "current" in use workspace and have it return some info about it, which then can be passed onto other resources like tfe_run_trigger

**Comment by joakimhellum** [comment]
Have created #207 as attempt to address this, which adds a `tfe_current_run` data source. However not sure if it's really appropriate for this provider :)

**Comment by chrisarcand** [comment]
Hi @joakimhellum! I've been doing some pondering about this problem space and while I don't yet have a solid path forward spelled out at the moment, I thought I'd share a bit of my findings and thoughts while it's still fresh in my head.

In short, the empty data source you wrote is clever and definitely not out of the question for the provider - but it creates an interesting coupling to the platform where the data source is very much dependent on additional ambient context that isn't available if you _aren't_ executing Terraform within Terraform Cloud/Enterprise. In other words, this datasource doesn't work when you're just using the provider without TFC, and it suddenly matters _where_ you are executing Terraform. This is also the case when you _are_ using TFC but using the [local execution mode](https://www.terraform.io/docs/cloud/workspaces/settings.html#execution-mode) on a workspace.

And that might be ok! Clearly no one executing something locally cares about this sort of datasource. But it got me thinking that perhaps we can just make fetching the data you need more ergonomic but still explicit (using normal variables provided by the TFC runtime environment, or provided by the user locally). There's a number of different ways you could do this, with various pros and cons.

TLDR: We're pondering about the best path forward on this issue and I'll keep you posted! As you mentioned above, the `tfe_current_run` may or may not be the most appropriate solution, but what it solves is definitely valid. 👍🏻

**Comment by joakimhellum** [comment]
@chrisarcand Many thanks for your feedback. Very much agree to not implement anything that creates more distance between Terraform local runs and Terraform Cloud runs :) 

We have been testing the `tfe_current_run` data source for a while now and have learned the following:

- We should not rely on the `tfe_current_run` data source for information about the repository. This was confusing because of what you aleready explained: the information is not available for local runs, even though the `.git` folder is. It's better to exclude `.git` folder in `.terraformignore` so it is uploaded to Terraform Cloud, then use one of the `git` providers in the registry to get information about the repository. 

- We didn't actually need the current workspace for our use case after further testing. In the end we just needed to get the OAuth Token ID from a list which we could get from another potential new data source: `tfe_oauth_tokens` / `tfe_oauth_clients`.

However I'm sure there are some valid reasons to get the current workspace and something similar could make it very convenient :)

Will go ahead and close #207 :)

**Comment by nemani** [comment]
Any update on this? 

We want to use terraform in a CI/CD pipeline, and want to create new workspaces and then modify the workspace settings (either change it to local execution mode from remote) or change and configure workspace env variables to setup aws creds on the newly created workspace. 

 I think if `tfe_current_run` is available I can make changes to the workspace before the rest of my module start using them, although I think this will be very tricky. 

Any idea on how I would be able to do this?

Thanks!

**Comment by chrisarcand** [comment]
Hi @nemani,

> We want to use terraform in a CI/CD pipeline, and want to create new workspaces and then modify the workspace settings (either change it to local execution mode from remote) or change and configure workspace env variables to setup aws creds on the newly created workspace.

Your specific situation may be more complex than this, but I will say there's nothing about the above that seems to indicate a need for the functionality discussed in this issue. You may do exactly this without any additional run context, today.

If you'd like more help on your specific issue and how to do something, the [HashiCorp Discuss forum](https://discuss.hashicorp.com/) is a better spot than here (This issue tracker is used for specific feature requests or bugs with the provider).

Thanks!

**Comment by chrisarcand** [comment]
> I'd like to associate the current in-use workspace to these created workspaces by way of tfe_run_trigger, but I cant see anyway to populate the value of the sourceable_id without hardcoded the current workspace name or id.

@A30004028 Circling back to your case, you can already do this today, [using the same mechanism that this suggested feature was to utilize](https://www.terraform.io/docs/cloud/run/run-environment.html#environment-variables):

```hcl
variable "TFC_WORKSPACE_NAME" {
  type = string
}

output "current_workspace_name" {
  value = var.TFC_WORKSPACE_NAME
}
```

```plaintext
Changes to Outputs:
  + current_workspace_name = "my-super-cool-workspace"
```

---

Circling back to this,`tfe_current_run` should definitely not be added, but you can do the exact same thing a different way:

An alternative that _does_ make sense and fulfills the same purpose you're looking for would be a general `tfe_run` data source. As `TFC_RUN_ID` is also provided to the run environment, you would then be able to use this datasource the exact same way as you're looking for with `tfe_current_run` (just, without the implicit environment requirement)

**Comment by jonmaestas** [comment]
@chrisarcand - Do those environment variables work for TFE installations as well? And - are these environment variables also included as TF variables? Based on your above example, it appears to be the case.

Here's the error I received with our TFE installation. 
```
The argument "TFC_WORKSPACE_SLUG" is required, but no definition was found.
```

I tried both `TFC_WORKSPACE_SLUG` and `TFE_WORKSPACE_SLUG` but no luck.

**Comment by jonmaestas** [comment]
@chrisarcand - I found it. TFE installations still uses `TFC_...` prefixes for these.

My above error was because it did not get passed into my module properly.

**Comment by joakimhellum** [comment]
Based on this discussion, the `tfe_current_run` data source should not be added, so will go ahead and close this issue :) For us this is very much question about managing "workspaces as code". Thanks.
